File: backend/src/main/java/com/meuprojeto/user/AuthController.java
package com.meuprojeto.user;

import com.meuprojeto.config.JwtUtil;
import com.meuprojeto.user.dto.*;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseCookie;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.web.bind.annotation.*;

/**
 * Autenticação baseada em cookies httpOnly:
 *  - access_token: JWT de vida curta, enviado em toda requisição autenticada.
 *  - refresh_token: token opaco de vida longa, usado só em /api/auth/refresh
 *    para obter um novo access_token sem exigir novo login. É rotacionado
 *    (revogado e substituído) a cada uso.
 *
 * Nenhum dos dois tokens é devolvido no corpo JSON, apenas via Set-Cookie —
 * isso limita o impacto de um eventual XSS no frontend, já que JS não
 * consegue ler cookies httpOnly.
 */
@RestController
@RequestMapping("/api/auth")
@RequiredArgsConstructor
public class AuthController {

    private static final String ACCESS_COOKIE = "access_token";
    private static final String REFRESH_COOKIE = "refresh_token";

    private final AuthenticationManager authenticationManager;
    private final UserService userService;
    private final JwtUtil jwtUtil;
    private final RefreshTokenService refreshTokenService;

    // Em produção, configure APP_COOKIE_SECURE=true (exige HTTPS) via variável de ambiente.
    @Value("${app.cookie-secure:false}")
    private boolean cookieSecure;

    @PostMapping("/register")
    public ResponseEntity<?> register(@Valid @RequestBody RegisterRequest request, HttpServletResponse response) {
        try {
            User user = userService.register(request);
            emitirCookies(user, response);
            return ResponseEntity.status(HttpStatus.CREATED).body(toAuthResponse(user));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.status(HttpStatus.CONFLICT).body(new ErrorResponse(e.getMessage()));
        }
    }

    @PostMapping("/login")
    public ResponseEntity<?> login(@Valid @RequestBody LoginRequest request, HttpServletResponse response) {
        try {
            authenticationManager.authenticate(
                    new UsernamePasswordAuthenticationToken(request.getEmail(), request.getSenha())
            );
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(new ErrorResponse("E-mail ou senha inválidos"));
        }

        User user = userService.findByEmail(request.getEmail());
        emitirCookies(user, response);
        return ResponseEntity.ok(toAuthResponse(user));
    }

    @PostMapping("/refresh")
    public ResponseEntity<?> refresh(@CookieValue(value = REFRESH_COOKIE, required = false) String refreshTokenValue,
                                      HttpServletResponse response) {
        if (refreshTokenValue == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(new ErrorResponse("Sessão expirada, faça login novamente"));
        }
        try {
            RefreshToken refreshToken = refreshTokenService.validar(refreshTokenValue);
            // Rotação: o refresh token usado é revogado e um novo par é emitido.
            refreshTokenService.revogar(refreshTokenValue);
            User user = refreshToken.getUsuario();
            emitirCookies(user, response);
            return ResponseEntity.ok(toAuthResponse(user));
        } catch (BadCredentialsException e) {
            limparCookies(response);
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(new ErrorResponse("Sessão expirada, faça login novamente"));
        }
    }

    @PostMapping("/logout")
    public ResponseEntity<Void> logout(@CookieValue(value = REFRESH_COOKIE, required = false) String refreshTokenValue,
                                        HttpServletResponse response) {
        if (refreshTokenValue != null) {
            refreshTokenService.revogar(refreshTokenValue);
        }
        limparCookies(response);
        return ResponseEntity.noContent().build();
    }

    private void emitirCookies(User user, HttpServletResponse response) {
        String accessToken = jwtUtil.generateToken(user);
        RefreshToken refreshToken = refreshTokenService.criar(user);

        ResponseCookie accessCookie = ResponseCookie.from(ACCESS_COOKIE, accessToken)
                .httpOnly(true)
                .secure(cookieSecure)
                .sameSite("Lax")
                .path("/")
                .maxAge(jwtUtil.getExpirationSeconds())
                .build();

        ResponseCookie refreshCookie = ResponseCookie.from(REFRESH_COOKIE, refreshToken.getToken())
                .httpOnly(true)
                .secure(cookieSecure)
                .sameSite("Lax")
                .path("/")
                .maxAge(refreshTokenService.getExpirationSeconds())
                .build();

        response.addHeader(HttpHeaders.SET_COOKIE, accessCookie.toString());
        response.addHeader(HttpHeaders.SET_COOKIE, refreshCookie.toString());
    }

    private void limparCookies(HttpServletResponse response) {
        ResponseCookie accessCookie = ResponseCookie.from(ACCESS_COOKIE, "")
                .httpOnly(true).secure(cookieSecure).sameSite("Lax").path("/").maxAge(0).build();
        ResponseCookie refreshCookie = ResponseCookie.from(REFRESH_COOKIE, "")
                .httpOnly(true).secure(cookieSecure).sameSite("Lax").path("/").maxAge(0).build();
        response.addHeader(HttpHeaders.SET_COOKIE, accessCookie.toString());
        response.addHeader(HttpHeaders.SET_COOKIE, refreshCookie.toString());
    }

    private AuthResponse toAuthResponse(User user) {
        return AuthResponse.builder()
                .id(user.getId())
                .nome(user.getNome())
                .email(user.getEmail())
                .role(user.getRole().name())
                .build();
    }

    record ErrorResponse(String message) {}
}

File: backend/src/auth/auth.controller.ts
import {
  Body,
  ConflictException,
  Controller,
  HttpCode,
  HttpStatus,
  Post,
  Req,
  Res,
  UnauthorizedException,
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';
import { ConfigService } from '@nestjs/config';
import { Request, Response } from 'express';
import { AuthService } from './auth.service';
import { UserService } from '../user/user.service';
import { JwtUtilService } from './jwt-util.service';
import { RefreshTokenService } from './refresh-token.service';
import { RegisterDto } from './dto/register.dto';
import { LoginDto } from './dto/login.dto';
import { Public } from '../common/decorators/public.decorator';
import { User } from '@prisma/client';

const ACCESS_COOKIE = 'access_token';
const REFRESH_COOKIE = 'refresh_token';

/**
 * Autenticação baseada em cookies httpOnly:
 *  - access_token: JWT de vida curta, enviado em toda requisição autenticada.
 *  - refresh_token: token opaco de vida longa, usado só em /api/auth/refresh
 *    para obter um novo access_token sem exigir novo login. É rotacionado
 *    (revogado e substituído) a cada uso.
 *
 * Nenhum dos dois tokens é devolvido no corpo JSON, apenas via Set-Cookie —
 * isso limita o impacto de um eventual XSS no frontend, já que JS não
 * consegue ler cookies httpOnly.
 */
@ApiTags('auth')
@Controller('api/auth')
export class AuthController {
  private readonly cookieSecure: boolean;

  constructor(
    private readonly authService: AuthService,
    private readonly userService: UserService,
    private readonly jwtUtil: JwtUtilService,
    private readonly refreshTokenService: RefreshTokenService,
    private readonly configService: ConfigService,
  ) {
    this.cookieSecure = this.configService.get<boolean>('app.cookieSecure', false);
  }

  @Public()
  @Post('register')
  async register(@Body() dto: RegisterDto, @Res({ passthrough: true }) response: Response) {
    try {
      const user = await this.userService.register(dto);
      await this.emitirCookies(user, response);
      response.status(HttpStatus.CREATED);
      return this.toAuthResponse(user);
    } catch (e) {
      if (e instanceof ConflictException) {
        response.status(HttpStatus.CONFLICT);
        return { message: e.message };
      }
      throw e;
    }
  }

  @Public()
  @HttpCode(HttpStatus.OK)
  @Post('login')
  async login(@Body() dto: LoginDto, @Res({ passthrough: true }) response: Response) {
    try {
      const user = await this.authService.validarCredenciais(dto.email, dto.senha);
      await this.emitirCookies(user, response);
      return this.toAuthResponse(user);
    } catch (e) {
      response.status(HttpStatus.UNAUTHORIZED);
      return { message: 'E-mail ou senha inválidos' };
    }
  }

  @Public()
  @HttpCode(HttpStatus.OK)
  @Post('refresh')
  async refresh(@Req() request: Request, @Res({ passthrough: true }) response: Response) {
    const refreshTokenValue = request.cookies?.[REFRESH_COOKIE];
    if (!refreshTokenValue) {
      response.status(HttpStatus.UNAUTHORIZED);
      return { message: 'Sessão expirada, faça login novamente' };
    }
    try {
      const refreshToken = await this.refreshTokenService.validar(refreshTokenValue);
      // Rotação: o refresh token usado é revogado e um novo par é emitido.
      await this.refreshTokenService.revogar(refreshTokenValue);
      await this.emitirCookies(refreshToken.usuario, response);
      return this.toAuthResponse(refreshToken.usuario);
    } catch (e) {
      if (e instanceof UnauthorizedException) {
        this.limparCookies(response);
        response.status(HttpStatus.UNAUTHORIZED);
        return { message: 'Sessão expirada, faça login novamente' };
      }
      throw e;
    }
  }

  @Public()
  @HttpCode(HttpStatus.NO_CONTENT)
  @Post('logout')
  async logout(@Req() request: Request, @Res({ passthrough: true }) response: Response) {
    const refreshTokenValue = request.cookies?.[REFRESH_COOKIE];
    if (refreshTokenValue) {
      await this.refreshTokenService.revogar(refreshTokenValue);
    }
    this.limparCookies(response);
  }

  private async emitirCookies(user: User, response: Response) {
    const accessToken = this.jwtUtil.generateToken(user);
    const refreshToken = await this.refreshTokenService.criar(user);

    response.cookie(ACCESS_COOKIE, accessToken, {
      httpOnly: true,
      secure: this.cookieSecure,
      sameSite: 'lax',
      path: '/',
      maxAge: this.jwtUtil.getExpirationSeconds() * 1000,
    });

    response.cookie(REFRESH_COOKIE, refreshToken.token, {
      httpOnly: true,
      secure: this.cookieSecure,
      sameSite: 'lax',
      path: '/',
      maxAge: this.refreshTokenService.getExpirationSeconds() * 1000,
    });
  }

  private limparCookies(response: Response) {
    response.cookie(ACCESS_COOKIE, '', {
      httpOnly: true,
      secure: this.cookieSecure,
      sameSite: 'lax',
      path: '/',
      maxAge: 0,
    });
    response.cookie(REFRESH_COOKIE, '', {
      httpOnly: true,
      secure: this.cookieSecure,
      sameSite: 'lax',
      path: '/',
      maxAge: 0,
    });
  }

  private toAuthResponse(user: User) {
    return {
      id: user.id,
      nome: user.nome,
      email: user.email,
      role: user.role,
    };
  }
}
